deeppavlov/models/elmo/elmo2tfhub.py

```diff
@@ -0,0 +1,217 @@
+# Copyright 2017 Neural Networks and Deep Learning lab, MIPT
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import tensorflow as tf
+import tensorflow_hub as hub
+import numpy as np
+
+from deeppavlov.models.elmo.elmo_model import BidirectionalLanguageModel, weight_layers
+
+
+def make_module_spec(options, weight_file):
+    """Makes a module spec.
+
+    Args:
+      options: LM hyperparameters.
+      weight_file: location of the hdf5 file with LM weights.
+
+    Returns:
+      A module spec object used for constructing a TF-Hub module.
+    """
+    def module_fn():
+        """Spec function for a token embedding module."""
+        # init
+        _bos_id = 256
+        _eos_id = 257
+        _bow_id = 258
+        _eow_id = 259
+        _pad_id = 260
+
+        _max_word_length = 50
+        _parallel_iterations = 10
+        _max_batch_size = 1024
+
+        id_dtype = tf.int32
+        id_nptype = np.int32
+        max_word_length = tf.constant(_max_word_length, dtype=id_dtype, name='max_word_length')
+
+        version = tf.constant('from_dp_1', dtype=tf.string, name='version')
+
+        # the charcter representation of the begin/end of sentence characters
+        def _make_bos_eos(c):
+            r = np.zeros([_max_word_length], dtype=id_nptype)
+            r[:] = _pad_id
+            r[0] = _bow_id
+            r[1] = c
+            r[2] = _eow_id
+            return tf.constant(r, dtype=id_dtype)
+
+        bos_ids = _make_bos_eos(_bos_id)
+        eos_ids = _make_bos_eos(_eos_id)
+
+        def token_level(token):
+            with tf.name_scope("token_level_preprocessor"):
+                char_ids = tf.decode_raw(token, tf.uint8, name='decode_raw2get_char_ids')
+                char_ids = tf.cast(char_ids, tf.int32, name='cast2int_token')
+                char_ids = tf.strided_slice(char_ids, [0], [max_word_length - 2],
+                                            [1], name='slice2resized_token')
+                ids_num = tf.shape(char_ids)[0]
+                fill_ids_num = (_max_word_length - 2) - ids_num
+                pads = tf.fill([fill_ids_num], _pad_id)
+                bow_token_eow_pads = tf.concat([[_bow_id], char_ids, [_eow_id], pads],
+                                               0, name='concat2bow_token_eow_pads')
+                return bow_token_eow_pads
+
+        def pre_sen_level(sen_dim):
```

Comment:
What does the name of this function mean? Maybe rename?

Reply:
Yes, it is a good idea